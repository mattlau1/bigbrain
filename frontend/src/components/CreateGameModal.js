import React, { useState } from 'react'
import { Button, Form, Modal } from 'react-bootstrap'
import API from '../utils/API';
import PropTypes from 'prop-types';
import { useAlert } from '../contexts/AlertProvider';

const CreateGameModal = ({ setShow, show, handleClose, handleShow, gameList, setGameList }) => {
  const [newGameName, setNewGameName] = useState('');
  const dispatch = useAlert();

  const createAlert = (type, message) => {
    dispatch({
      type: type,
      message: message,
    })
  }

  const createGame = async (e) => {
    e.preventDefault()
    if (!newGameName) {
      createAlert('ERROR', 'Your new quiz cannot have an empty name');
      return;
    }

    const body = {
      name: newGameName,
    };

    try {
      const token = localStorage.getItem('token');
      const api = new API();
      const res = await api.postAPIRequestBodyToken('admin/quiz/new', body, token);
      if (res.ok) {
        handleClose();
        createAlert('SUCCESS', 'Successfully created a new quiz!');

        api.getAPIRequestToken('admin/quiz', token).then((data) => {
          if (data.status === 403) {
            createAlert('ERROR', 'Invalid Token');
          } else if (data.status === 200) {
            setGameList([]);
            data.json().then((quizzes) => {
              quizzes.quizzes.forEach((quiz) => {
                api.getAPIRequestToken(`admin/quiz/${quiz.id}`, token).then((data) => {
                  data.json().then((quizData) => {
                    const newGame = { ...quizData, ...quiz }
                    setGameList(gameList => [...gameList, newGame]);
                  })
                }).catch((e) => {
                  createAlert('ERROR', 'There was a problem getting quizzes');
                  console.warn(e)
                })
              })
            })
          }
        }).catch((e) => {
          createAlert('ERROR', 'There was a problem getting quizzes');
          console.warn(e)
        })
      } else {
        createAlert('ERROR', 'There was a problem creating a new quiz');
      }
    } catch (e) {
      createAlert('ERROR', 'An unexpected error has occured');
      console.warn(e);
    }
  }

  return (
        <Modal show={show} onHide={handleClose}>
        <Form>
            <Modal.Header closeButton>
            <Modal.Title>Create Game</Modal.Title>
            </Modal.Header>
            <Modal.Body>
            <Form.Control
                className="inputBox"
                type="text"
                placeholder="Game Title"
                onChange={(e) => setNewGameName(e.target.value)}
                id='textInput'
            />
            </Modal.Body>
            <Modal.Footer>
            <Button id='closeButton' variant="secondary" onClick={handleClose}>
                Close
            </Button>
            <Button id='createButton' variant="primary" type="submit" onClick={(e) => { createGame(e) } }>
                Create Game
            </Button>
            </Modal.Footer>
        </Form>
      </Modal>
  )
}

CreateGameModal.propTypes = {
  setShow: PropTypes.func,
  handleShow: PropTypes.func,
  handleClose: PropTypes.func,
  show: PropTypes.bool,
  gameList: PropTypes.array,
  setGameList: PropTypes.func
};

export default CreateGameModal
