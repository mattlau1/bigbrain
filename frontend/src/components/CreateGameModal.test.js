import { shallow } from 'enzyme';
import React from 'react';
import CreateGameModal from './CreateGameModal';
import renderer from 'react-test-renderer';

describe('Create Game Modal', () => {
  it('should render create game modal component', () => {
    const tree = renderer.create(<CreateGameModal />).toJSON();
    const wrapper = shallow(<CreateGameModal />);
    expect(wrapper).toHaveLength(1);

    // ensure that there are no changes to the component after test
    // (before/after is the same)
    expect(tree).toMatchSnapshot();
  })

  it('should have the close button', () => {
    const tree = renderer.create(<CreateGameModal />).toJSON();
    const wrapper = shallow(<CreateGameModal />);
    const text = wrapper.find('#close-button');
    expect(text.text()).toBe('Close');

    // ensure that there are no changes to the component after test
    // (before/after is the same)
    expect(tree).toMatchSnapshot();
  })

  it('should have the create game button', () => {
    const tree = renderer.create(<CreateGameModal />).toJSON();
    const wrapper = shallow(<CreateGameModal />);
    const text = wrapper.find('#create-button');
    expect(text.text()).toBe('Create Game');

    // ensure that there are no changes to the component after test
    // (before/after is the same)
    expect(tree).toMatchSnapshot();
  })

  it('should have the text input box', () => {
    const tree = renderer.create(<CreateGameModal />).toJSON();
    const wrapper = shallow(<CreateGameModal />);
    expect(wrapper.find('#game-title-input')).toHaveLength(1);

    // ensure that there are no changes to the component after test
    // (before/after is the same)
    expect(tree).toMatchSnapshot();
  })

  it('should have any text if modal is not on display', () => {
    const tree = renderer.create(<CreateGameModal />).toJSON();
    const wrapper = shallow(<CreateGameModal />);
    wrapper.find('#game-title-input');
    wrapper.simulate('change', { target: { value: 'A new game' } })
    const text = wrapper.find('#game-title-input');
    expect(text.text()).toBe('');

    // ensure that there are no changes to the component after test
    // (before/after is the same)
    expect(tree).toMatchSnapshot();
  })
})
