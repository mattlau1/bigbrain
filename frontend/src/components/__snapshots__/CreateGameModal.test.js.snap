// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Game Modal should have any text if modal is not on display 1`] = `null`;

exports[`Create Game Modal should have the close button 1`] = `null`;

exports[`Create Game Modal should have the create game button 1`] = `null`;

exports[`Create Game Modal should have the text input box 1`] = `null`;

exports[`Create Game Modal should render create game modal component 1`] = `null`;
